Fix #3300: Add check for accessibility label for activities

Sparsh1212 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes #3300: Add check for accessibility labels for activities

The script parses the `AndroidManifest.xml` file of the `app` layer to ensure that the `activity` elements are defined with accessibility labels.

**Note:** The Activities listed in the `ACCESSIBILITY_LABEL_CHECK_EXEMPTIONS_LIST` will be skipped for this check, and currently, all the activities which were failing this check, have been added to this list so that the workflow doesn't fail.
We can remove any Activity from the exemption list, where we think that we can't go without a label for that Activity, or if a label is added for that Activity.

To run the script, use:

`bazel run //scripts:accessibility_label_check -- $(pwd) app/src/main/AndroidManifest.xml`

For testing the script, automated tests have been added.
To execute the tests, use:

`bazel test //scripts:AccessibilityLabelCheckTest`

**Note:** We are generating the test assets dynamically at the time of executing them. The test assets are automatically deleted when the test finishes.

SS when the CI check fails:

![image](https://user-images.githubusercontent.com/55937724/125578039-f9d3f1fc-ddf4-4694-9ef0-61ed8ca4423d.png)


## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sparsh1212 committed:
Fix #3290: Add support for generic regex pattern matching

Sparsh1212 committed:
Apply review suggestions for proto location

Sparsh1212 committed:
Delete main directory

Sparsh1212 committed:
Implment script logic and tests

Sparsh1212 committed:
CI setup part 1

Sparsh1212 committed:
syntax fix in yaml file

Sparsh1212 committed:
import nits in dummy file

Sparsh1212 committed:
Use regex patterns from script in script test

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Improve naming of script_assets variables

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Revamp testing approach

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Fix #3291: Add check for XML syntax correctness

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Introduce XML syntax check in static checks workflow

Sparsh1212 committed:
Fix #3292: Add check for test files presence for prod files

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Fix #3300: Add check for accessibility labels for activities

Sparsh1212 committed:
Add CI check for label presence for activities

fsharpasharp reviewed (approved):
Approving for BUILD.bazel change

Sparsh1212 committed:
Add ScriptConstants file to exemptions test list

Sparsh1212 committed:
merge test-file-check

Sparsh1212 committed:
Test if: always()

Sparsh1212 committed:
Revert "Test if: always()"

This reverts commit b72e419b

Sparsh1212 committed:
Add if: always()

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Add if: always()

Sparsh1212 committed:
Merge branch 'test-file-check' into accesibilty-label-check

Sparsh1212 committed:
Add if: always()

anandwana001 commented:
@Sparsh1212 it looks like after assigning, you are making some commits, 
I suggest following the below guide to maintain the assign review process.
https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section

Sparsh1212 committed:
Apply review suggestions on PR

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make review suggestions from regex pattern checks

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Implement review suggestions based on #3340

anandwana001 reviewed (changes requested):
Suggesting to follow the pr review suggestions from #3340

Sparsh1212 committed:
Refactor PR as per feedback recieved

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Implement feedback suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nits

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
merge test-file-check

Sparsh1212 committed:
Implement feedback suggestions as per #3340

Sparsh1212 commented:
> Suggesting to follow the pr review suggestions from #3340

Done! PTAL.

Sparsh1212 commented:
@BenHenning Updated this PR as per your initial feedback received on #3340. PTAL.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,108 @@
+import java.util.stream.IntStream
+import javax.xml.parsers.DocumentBuilderFactory
+
+class AccessibilityLabelCheck {
```
kdocs

Sparsh1212 replied:
Done.

rt4914 replied:
Resolving after confirming this has been done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  /**
+   * The [NodeList] is not iterable. This helper
```
any time you want to mention any argument in comment, put argument name, not the argument type 

```suggestion
   * The [nodeList] is not iterable. This helper
```

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/ExemptionsList.kt:
```diff
@@ -2,6 +2,35 @@ package org.oppia.android.scripts
 
 /** Lists of files which should be exepmted for the different script checks. */
 object ExemptionsList {
+  val ACCESSIBILITY_LABEL_CHECK_EXEMPTIONS_LIST = arrayOf(
```
kdoc

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/AccessibilityLabelCheckTest.kt:
```diff
@@ -0,0 +1,195 @@
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+
+class AccessibilityLabelCheckTest {
```
kdoc

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/AccessibilityLabelCheckTest.kt:
```diff
@@ -0,0 +1,195 @@
+    manifestFile.writeText(manifestContent)
+
+    val exception = assertThrows(Exception::class) {
+      runScript("AndroidManifest.xml")
```
named argument

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/AccessibilityLabelCheckTest.kt:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  @Test
+  fun testAccessibilityLabel_labelNotPresent_activityAreNotDefinedWithAccessibilityLabels() {
```
`activityAreNotDefinedWithAccessibilityLabels`

`activityAre` - Update based on what we are testing, singular or plural.

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/AccessibilityLabelCheckTest.kt:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  @Test
+  fun testAccessibilityLabel_labelPresent_activitiesAreDefinedWithAccessibilityLabels() {
```
also write positive test, that all activities have labels.

Sparsh1212 replied:
Apologies! I didn't catch this correctly. This test actually does the same thing which you suggested. Here all the activities are defined with the label, and we are asserting that the script check is successful here. Please let me know if this isn't still clear.

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Merge branch 'test-file-check' into accesibilty-label-check

Sparsh1212 committed:
nit fix

vinitamurthi reviewed (approved):
LGTM for codeowner files

anandwana001 reviewed (approved):
LGTM,

Please address all the suggested reviews.

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Merge branch 'test-file-check' into accesibilty-label-check

rt4914 commented on scripts/src/java/org/oppia/android/scripts/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,112 @@
+      doc.getDocumentElement().normalize()
+
+      // list of all the activity elements.
+      val nodeList = accessibilityLabelCheck.convertNodeListToListOfNode(
```
Name this val as `completeActivityList`

Sparsh1212 replied:
Done.

rt4914 commented on scripts/src/java/org/oppia/android/scripts/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,112 @@
+      )
+
+      // list of activity elements which lacks the label.
+      val matchedNodes = nodeList.filter { node ->
```
Name this val as `activityListWithoutLabel`

Sparsh1212 replied:
Done.

rt4914 commented on scripts/src/java/org/oppia/android/scripts/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,112 @@
+  companion object {
+    @JvmStatic
+    fun main(vararg args: String) {
+      // path of the repo to be analyzed.
```
I think internal comments are not needed as the `val` names are already clear.

Sparsh1212 replied:
Removed.

rt4914 commented on scripts/src/java/org/oppia/android/scripts/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,112 @@
+   * @return a list of nodes
+   */
+  private fun convertNodeListToListOfNode(nodeList: NodeList): List<Node> {
+    return IntStream.range(
```
Suggestion:
Write it as below as it becomes more readable
```
    return IntStream.range(0, nodeList.getLength())
      .mapToObj(nodeList::item)
      .collect(Collectors.toList())
```

Sparsh1212 replied:
Done.

rt4914 commented on scripts/src/java/org/oppia/android/scripts/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,112 @@
+  private fun checkIfActivityHasMissingLabel(node: Node): Boolean {
+    val attributesList = node.getAttributes()
+    val activityPath = attributesList.getNamedItem("android:name").getNodeValue()
+    if (
```
Remove this entire `if` and simply write
```
    return activityPath !in ExemptionsList.ACCESSIBILITY_LABEL_CHECK_EXEMPTIONS_LIST && 
      attributesList.getNamedItem("android:label") == null
```

Sparsh1212 replied:
Done, thanks.

rt4914 reviewed (changes requested):
@Sparsh1212 Nit changed suggested. Thanks.

Sparsh1212 committed:
nit fix

Sparsh1212 commented:
> @Sparsh1212 Nit changed suggested. Thanks.

Thanks!
PR updated. Please check.

BenHenning commented:
Skipping this review for now since this is part of milestone 2.

rt4914 commented on scripts/src/java/org/oppia/android/scripts/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  /**
+   * Checks whether a activity element has a missing label.
```
`an` instead of `a`

`Checks whether an activity element has a missing label.`

Sparsh1212 replied:
Done

Sparsh1212 replied:
@rt4914 Can we resolve this now?

rt4914 reviewed (approved):
@Sparsh1212 One nit suggested.

Do not merge as primary mentor is Ben. Wait for his review later.

rt4914 reviewed (changes requested):
Blocking for @BenHenning 's review which will be done in Milestone 2.

Sparsh1212 committed:
Review suggestions part 1

Sparsh1212 committed:
Implement review suggestions part 2

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
resolve merge conflict

Sparsh1212 committed:
update static_checks

Sparsh1212 committed:
add test to repository file

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
Do review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Add Ben as a codeowner for the ScriptExemptions file

Sparsh1212 committed:
bazel files nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Introduce a library for the regex assets

BenHenning commented:
@Sparsh1212 please assign back to me once your milestone 1 PRs are all at least approved.

Sparsh1212 committed:
Make directory structure same as that of #3374

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make directory structure similar to #3374

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
merge xml-syntax-checks

Sparsh1212 committed:
add new files to test file script exemptions and nit fixes

Sparsh1212 committed:
merge develop

Sparsh1212 committed:
diable ktlint max-line-length

Sparsh1212 committed:
disable ktlint-max-line

Sparsh1212 committed:
disable ktlint max-length

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
remove testonly attribute from tests

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add new files to script exemptions

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Apply review suggestions

Sparsh1212 committed:
ktlint fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Apply nit fixes

Sparsh1212 committed:
Apply nit fixes

Sparsh1212 committed:
Remove script constants

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
add tests for XML syntax error handler

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 committed:
add todo

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add todo

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
add test

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add testOnly

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
fix textproto file

Sparsh1212 committed:
merge-regex-checks

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 committed:
add exemptions to textproto

Sparsh1212 committed:
sort failures lexicographically

Sparsh1212 committed:
Merge branch 'develop' into test-file-check

Sparsh1212 committed:
add new files to exemption list

Sparsh1212 committed:
merge test-file-check

Sparsh1212 committed:
merge develop

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
add kdocs

Sparsh1212 commented:
@BenHenning @rt4914 PTAL.

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

BenHenning commented:
Hi @Sparsh1212. I plan to take a look at this tomorrow.

Please make sure the PR description is updated to:
- Not include blocked information since that's no longer relevant
- Include an example of the CI check when it legitimately fails (for context on what the error messaging looks like)

Please reassign when done.

Sparsh1212 committed:
Trigger CI failure exclusively

Sparsh1212 commented:
> Hi @Sparsh1212. I plan to take a look at this tomorrow.
> 
> Please make sure the PR description is updated to:
> 
> * Not include blocked information since that's no longer relevant
> * Include an example of the CI check when it legitimately fails (for context on what the error messaging looks like)
> 
> Please reassign when done.

Done.

Sparsh1212 committed:
Revert CI check failure

rt4914 reviewed (approved):
LGTM, thanks.

BenHenning commented:
Probably will need to look at this later yet--sorry about that (had a lot come up this week).

Sparsh1212 committed:
merge develop

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

BenHenning commented on scripts/assets/accessibility_label_exemptions.textproto:
```diff
@@ -0,0 +1,27 @@
+exempted_activity: "app/home/HomeActivity"
+exempted_activity: "app/home/recentlyplayed/RecentlyPlayedActivity"
+exempted_activity: "app/mydownloads/MyDownloadsActivity"
+exempted_activity: "app/player/state/testing/StateFragmentTestActivity"
+exempted_activity: "app/profileprogress/ProfilePictureActivity"
+exempted_activity: "app/settings/profile/ProfileEditActivity"
+exempted_activity: "app/splash/SplashActivity"
+exempted_activity: "app/testing/AudioFragmentTestActivity"
+exempted_activity: "app/testing/BindableAdapterTestActivity"
+exempted_activity: "app/testing/ConceptCardFragmentTestActivity"
+exempted_activity: "app/testing/DragDropTestActivity"
+exempted_activity: "app/testing/ExplorationInjectionActivity"
+exempted_activity: "app/testing/ExplorationTestActivity"
+exempted_activity: "app/testing/HomeFragmentTestActivity"
+exempted_activity: "app/testing/HomeTestActivity"
+exempted_activity: "app/testing/HtmlParserTestActivity"
+exempted_activity: "app/testing/ImageRegionSelectionTestActivity"
+exempted_activity: "app/testing/InputInteractionViewTestActivity"
+exempted_activity: "app/testing/LessonThumbnailImageViewTestActivity"
+exempted_activity: "app/testing/NavigationDrawerTestActivity"
+exempted_activity: "app/testing/ProfileChooserFragmentTestActivity"
+exempted_activity: "app/testing/SplashTestActivity"
+exempted_activity: "app/testing/TestFontScaleConfigurationUtilActivity"
+exempted_activity: "app/testing/TopicRevisionTestActivity"
+exempted_activity: "app/testing/TopicTestActivity"
+exempted_activity: "app/testing/TopicTestActivityForStory"
+exempted_activity: "app/walkthrough/WalkthroughActivity"
```
@rt4914 do we already have issues tracking these?

rt4914 replied:
Not for all.
For example:
`HomeActivity` does not need a label as it works correctly even without it for A11y users.
`NavigationDrawerTestActivity ` we don't need labels for them.


So what are your thoughts on both of these? Should we still add labels for all the files?

BenHenning replied:
I think we can ignore the test activities--they're reasonable exemptions (eventually we'll probably delete them by utilizing a generic replacement, anyway).

For the others, shouldn't we always have an accessibility label to be explicit? If so, then we ought to have a TODO for each of the non-test activities to ensure we can remove them from this exemption list & only leave test activities.

WDYT?

rt4914 replied:
Yes your suggestion makes sense.

Filed an issue: https://github.com/oppia/oppia-android/issues/3602

BenHenning replied:
Thanks @rt4914!

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+ *
+ * Arguments:
+ * - path_to_directory_root: directory path to the root of the Oppia Android repository.
+ * - path_to_app_level_manifest_file: path to the manifest file of app layer.
```
Rather than 'app-level manifest', I suggest framing it as the 'main manifest of the app' since there won't be an "app layer" or "app module" in the future.

That being said, I think we should actually make this generic & allow any number of manifests to be passed in (& test multiple in the tests). Longer term, we'll be moving away from a single manifest to multiple that are merged together using the Bazel Android manifest merging. This actually lets us properly exclude test activities from the production build that we ship to users. Rather than needing to update this check then, it'd be great if we could just support it in the initial version. Do you have any concerns with this?

Sparsh1212 replied:
Revamped the approach to make it generic, now we can pass multiple manifests.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+    accessibilityLabelExemptiontextProto
+  ).getExemptedActivityList()
+
+  val manifesFilePath = args[1]
```
Nit: fix typo in variable name.

Sparsh1212 replied:
Outdated.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+fun main(vararg args: String) {
+  val repoPath = "${args[0]}/"
+
+  val accessibilityLabelExemptiontextProto = "scripts/assets/accessibility_label_exemptions"
```
Ending should be ``TextProto`` (as separate words in the identifier).

Ditto in the function below with the parameter of the same name.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+ * @param accessibilityLabelExemptionList list of all the exemptions of the label check
+ * @return whether the label is present or not
+ */
+private fun checkIfActivityHasMissingLabel(
```
The name is a bit grammatically off. Maybe: ``checkIfActivityIsMissingAccessibilityLabel`` or ``isActivityMissingAccessibilityLabel``?

Sparsh1212 replied:
This is outdated.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+  val attributesList = activityNode.getAttributes()
+  val activityPath = attributesList.getNamedItem("android:name").getNodeValue()
+  if (
+    activityPath
+      .removePrefix(".")
+      .replace(".", "/") in accessibilityLabelExemptionList
```
I suggest extracting the 'removePrefix.replace()' transformation to a separate variable so that the if-check can fit on one line (for better readability/code clarity).

Sparsh1212 replied:
This is outdated.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+}
+
+/**
+ * The [nodeList] is not iterable. This helper
+ * function converts it to List<Node>.
```
Please only wrap KDoc lines when they reach the 100 character boundary, or when a new paragraph is needed (in which case an empty line should separate the paragraphs).

Sparsh1212 replied:
This is outdated.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+ * @param nodeList instance of NodeList
+ * @return a list of nodes
+ */
+private fun convertNodeListToListOfNode(nodeList: NodeList): List<Node> {
+  return IntStream.range(0, nodeList.getLength())
+    .mapToObj(nodeList::item)
+    .collect(Collectors.toList())
+}
```
We could use an extension function here:

```suggestion
private fun NodeList.toListOfNodes(): List<Node> {
  return IntStream.range(0, getLength()).mapToObj(this::item).collect(Collectors.toList())
}
```

Also, a more Kotlin idiomatic approach would probably be to use an IntRange instead:

```suggestion
private fun NodeList.toListOfNodes(): List<Node> = (0 until getLength()).map(this::item)
```

Sparsh1212 replied:
This looks neat. Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+      println(
+        "- $pathPrefix" +
+          "${
+          node.getAttributes()
+            .getNamedItem("android:name")
+            .getNodeValue()
+            .replace(".", "/")
```
Please de-duplicate with the code above doing the same thing (I suggest using a data class or helper to avoid needing duplicate the code in both places).

Sparsh1212 replied:
This is outdated.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+ *     textproto file.
+ * @return proto class from the parsed textproto file
+ */
+private fun loadAccessibilityLabelExemptionsProto(accessibilityLabelExemptiontextProto: String):
+  AccessibilityLabelExemptions {
+    val protoBinaryFile = File("$accessibilityLabelExemptiontextProto.pb")
+    val builder = AccessibilityLabelExemptions.getDefaultInstance().newBuilderForType()
+
+    // This cast is type-safe since proto guarantees type consistency from mergeFrom(),
+    // and this method is bounded by the generic type T.
+    @Suppress("UNCHECKED_CAST")
+    val protoObj: AccessibilityLabelExemptions =
+      FileInputStream(protoBinaryFile).use {
+        builder.mergeFrom(it)
+      }.build() as AccessibilityLabelExemptions
+    return protoObj
+  }
```
```suggestion
private fun loadAccessibilityLabelExemptionsProto(
  accessibilityLabelExemptiontextProto: String
): AccessibilityLabelExemptions {
  val protoBinaryFile = File("$accessibilityLabelExemptiontextProto.pb")
  val builder = AccessibilityLabelExemptions.getDefaultInstance().newBuilderForType()
   // This cast is type-safe since proto guarantees type consistency from mergeFrom(),
  // and this method is bounded by the generic type T.
  @Suppress("UNCHECKED_CAST")
  val protoObj: AccessibilityLabelExemptions =
    FileInputStream(protoBinaryFile).use {
      builder.mergeFrom(it)
    }.build() as AccessibilityLabelExemptions
  return protoObj
}
```

Prefer breaking at the arguments to avoid needing to indent the entire function body.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/proto/BUILD.bazel:
```diff
@@ -62,3 +62,15 @@ java_lite_proto_library(
+
+proto_library(
+    name = "accessibility_label_exemptions_proto",
+    srcs = ["script_exemptions.proto"],
```
I suggest renaming the existing file & creating a new one--we shouldn't be reusing the same .proto file in multiple libraries. That leads to duplicate build graph work, and potentially duplicate symbols/single version violations in the graph which can actually cause runtime issues in the worst cases.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/label/AccessibilityLabelCheckTest.kt:
```diff
@@ -0,0 +1,353 @@
+  fun testAccessibilityLabel_labelPresent_activitiesAreDefinedWithAccessibilityLabel() {
+    val manifestContent =
+      """
+      <?xml version="1.0" encoding="utf-8"?>
+      <manifest xmlns:android="http://schemas.android.com/apk/res/android"
+        xmlns:tools="http://schemas.android.com/tools"
+        package="org.oppia.android">
+
+        <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
+        <uses-permission android:name="android.permission.INTERNET" />
+        <application
+          android:name=".app.application.OppiaApplication"
+          android:allowBackup="true"
+          android:icon="@mipmap/ic_launcher"
+          android:label="@string/app_name"
+          android:roundIcon="@mipmap/ic_launcher_round"
+          android:supportsRtl="true"
+          android:theme="@style/OppiaTheme">
+          <meta-data
+            android:name="firebase_analytics_collection_deactivated"
+            android:value="true" />
+          <meta-data
+            android:name="firebase_crashlytics_collection_enabled"
+            android:value="false" />
+          <meta-data
+            android:name="automatic_app_expiration_enabled"
+            android:value="false" />
+          <meta-data
+            android:name="expiration_date"
+            android:value="2020-09-01" />
```
I suggest removing all the cruft that's not actually needed for the purpose of these tests.

Sparsh1212 replied:
Removed.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/label/BUILD.bazel:
```diff
@@ -0,0 +1,18 @@
+"""
+Tests corresponding to Accessibility label check that ensures all the activities in the codebase
```
```suggestion
Tests corresponding to accessibility label check that ensures all the activities in the codebase
```

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,156 @@
+  val activityPath = attributesList.getNamedItem("android:name").getNodeValue()
+  if (
+    activityPath
+      .removePrefix(".")
```
Is this sufficient? I believe the name can also be an absolute, rather than relative, path. Also, shouldn't we be taking into consideration the path of the manifest file to ensure the fully qualified activity package can be computed? For example, a manifest under ``app/src/main/java/org/oppia/android/splash`` could just reference '.SplashActivity' since it's in the same package.

Sparsh1212 replied:
The approach is now revamped. Now for every activity, we compute the fully qualified package name, by parsing the manifest file and obtaining the value from the `package` attribute. Through this approach, the example you stated above works fine, also added tests for the same for verification.

BenHenning reviewed (changes requested):
Thanks @Sparsh1212! Took a first pass. I didn't look at the tests in detail yet, but I will after this batch of comments are addressed.

Sparsh1212 committed:
nit fixes

Sparsh1212 commented:
> Thanks @Sparsh1212! Took a first pass. I didn't look at the tests in detail yet, but I will after this batch of comments are addressed.

The comments are now addressed. PTAL.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -5,126 +5,145 @@ import org.w3c.dom.Node
  * Usage:
- *   bazel run //scripts:accessibility_label_check -- <path_to_directory_root> <path_to_app_level
- *   _manifest_file>
+ *   bazel run //scripts:accessibility_label_check -- <path_to_directory_root> <path_to_manifest
```
```suggestion
 *   bazel run //scripts:accessibility_label_check -- <path_to_directory_root> <paths_to_manifest
```

However, it should be something like:

``[paths to manifest files...]`` to indicate that it's space-separated.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -5,126 +5,145 @@ import org.w3c.dom.Node
+    val file = File(repoPath + relativePath)
+    val docBuilder = builderFactory.newDocumentBuilder()
+    val doc = docBuilder.parse(file)
+    doc.getDocumentElement().normalize()
```
Why is this needed? I suggest adding a brief line comment to provide context.

Sparsh1212 replied:
Added.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -5,126 +5,145 @@ import org.w3c.dom.Node
+  if (activityName.startsWith(".")) {
+    return activityPathPrefix.plus(packageName.plus(activityName).replace(".", "/"))
+  } else {
+    return activityPathPrefix.plus(activityName.replace(".", "/"))
```
It seems a bit odd to call plus directly. Could we just use the ``+`` operator instead?

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -5,126 +5,145 @@ import org.w3c.dom.Node
+): String {
+  if (activityName.startsWith(".")) {
+    return activityPathPrefix.plus(packageName.plus(activityName).replace(".", "/"))
+  } else {
```
Are both of these cases covered in tests?

Sparsh1212 replied:
Yes.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -5,126 +5,145 @@ import org.w3c.dom.Node
+    val doc = docBuilder.parse(file)
+    doc.getDocumentElement().normalize()
+    val packageName = doc.getDocumentElement().getAttribute("package")
+    doc.getElementsByTagName("activity").toListOfNodes().mapNotNull { activityNode ->
```
```suggestion
    return@flatMap doc.getElementsByTagName("activity").toListOfNodes().mapNotNull { activityNode ->
```

Optional, but I find the extra ``return`` can help with clarity for longer lambdas.

Sparsh1212 replied:
Added.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -5,126 +5,145 @@ import org.w3c.dom.Node
   val repoPath = "${args[0]}/"
 
-  val accessibilityLabelExemptiontextProto = "scripts/assets/accessibility_label_exemptions"
+  val accessibilityLabelExemptionTextProto = "scripts/assets/accessibility_label_exemptions"
```
```suggestion
  val accessibilityLabelExemptionTextProtoFilePath = "scripts/assets/accessibility_label_exemptions"
```

Ditto elsewhere this same name is used. The name wasn't indicating what it contained before--I missed this in the last pass.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -5,126 +5,145 @@ import org.w3c.dom.Node
+  if (missingAccessibilityLabelActivities.isNotEmpty()) {
+    println("Accessibility label missing for Activities:")
+    missingAccessibilityLabelActivities.sorted().forEach { activityPath ->
+      println("- ${repoPath.plus(activityPath)}")
```
Ditto here for 'plus'. I also assume you could just do ``"- $repoPath$activityPath"`` instead.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/label/AccessibilityLabelCheck.kt:
```diff
@@ -0,0 +1,176 @@
+  val builderFactory = DocumentBuilderFactory.newInstance()
+
+  val missingAccessibilityLabelActivities = manifestPaths.flatMap { relativePath ->
+    val file = File(repoPath + relativePath)
```
We probably shouldn't be assuming repoPath ends in a path separator. Suggest instead doing something like:

```kotlin
val repoRoot = File(repoPath)
val missingAccessibilityLabelActivities = ... {
  val file = File(repoRoot, relativePath)
  ...
```

To improve the script's robustness.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/label/AccessibilityLabelCheckTest.kt:
```diff
@@ -0,0 +1,300 @@
+  }
+
+  @Test
+  fun testAccessibilityLabel_AccessibilitylabelNotDefinedForExemptedActivity_checkShouldPass() {
```
```suggestion
  fun testAccessibilityLabel_accessibilityLabelNotDefinedForExemptedActivity_checkShouldPass() {
```

Note capitalization.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/label/AccessibilityLabelCheckTest.kt:
```diff
@@ -0,0 +1,300 @@
+    val failureMessage =
+      """
+      Accessibility label missing for Activities:
+      - ${retrieveTestFilesDirectoryPath()}/$firstAppActivityPath
```
Do you think it might be worth outputting the relative paths such that copying them into the exemptions file will just work?

Sparsh1212 replied:
Done.

BenHenning reviewed (commented):
Thanks @Sparsh1212! Generally LGTM, but had a few more comments. PTAL.

Sparsh1212 committed:
Merge branch 'develop' into accesibilty-label-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Add redundant exemptions check

Sparsh1212 committed:
Merge branch 'develop' into accesibilty-label-check

Sparsh1212 committed:
rectify static_checks

Sparsh1212 committed:
Remove duplicate proto libraries

Sparsh1212 commented:
@BenHenning 
- All comments addressed
- Migrated multiple exemption protos to share a single proto file and library
- Added a check for redundant exemptions
- Added tests for redundant exemptions

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/label/AccessibilityLabelCheckTest.kt:
```diff
@@ -285,14 +356,111 @@ class AccessibilityLabelCheckTest {
+    val manifestFile = tempFolder.newFile("testfiles/$tempFileRelativePath")
+    manifestFile.writeText(testContent)
+    val exemptionFile = File("${tempFolder.root}/$pathToProtoBinary")
+    val builder = AccessibilityLabelExemptions.newBuilder()
+    builder.addExemptedActivity("app/src/main/java/org/oppia/android/app/home/HomeActivity")
+    val exemptions = builder.build()
```
Here & elsewhere: consider using the ``apply{}`` pattern to construct the proto in one statement.

Sparsh1212 replied:
Done.

BenHenning reviewed (approved):
Thanks @Sparsh1212. Just had 2 comments, otherwise this LGTM.

BenHenning reviewed (changes requested):
Marking as requested to avoid this PR being accidentally merged before the last 2 comments are resolved.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/label/AccessibilityLabelCheckTest.kt:
```diff
@@ -0,0 +1,468 @@
+      <manifest xmlns:android="http://schemas.android.com/apk/res/android"
+          package="org.oppia.android">
+          <activity
+              android:name=".app.home.SplashActivity" />
```
Can you also add a test where the activity's reference is fully qualified rather than relative?

Sparsh1212 replied:
This was already added, renamed it now to give a better context.

Sparsh1212 committed:
nit fixes

Sparsh1212 commented:
@BenHenning Thanks!
Changes done. PTAL.

Sparsh1212 committed:
add activity to exemption

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'develop' into accesibilty-label-check

Sparsh1212 committed:
update branch

BenHenning reviewed (approved):
Thanks @Sparsh1212! LGTM.
